### PR TITLE
Revert "Revert "[core][scalability] Offload RPC Finish to a thread pool.""

### PR DESCRIPTION
**Pull request by fishbone**
Reverts ray-project/ray#30737

### CODE REVIEW

**Comment by iycheng** [comment]
The only change is: https://github.com/ray-project/ray/pull/30766/files/f72ce8f467d039aa34805c59f0a3987b1b0b7bd6..893c05b91cfbd9872a41e62f606bbe1cc90396fa

**Comment by iycheng** [comment]
Running test here:
https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_rhdfm4es3cy7cjqxrpf3rmxr
https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_rmjr4ysld7u28v84ywz16g6m
https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_wixt8a1pc3x72bzu4n9vj5ut
https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_vny6pic518euzr9zse23fg8m

**Comment by iycheng** [comment]
All passed
```
  wheels_url:    https://ray-ci-artifact-pr-public.s3.amazonaws.com/893c05b91cfbd9872a41e62f606bbe1cc90396fa/tmp/artifacts/.whl/ray-3.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
  cluster_url:   https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_wixt8a1pc3x72bzu4n9vj5ut

[INFO 2022-12-03 08:01:54,929] log.py: 36  Observed the following results:

  iteration = 2739
  iteration_time = 72.74934101104736
  absolute_time = 1670054482.16392
  elapsed_time = 86393.33542346954
  last_update = 1670054482.163943
  last_update_diff = 31.209413766860962

[INFO 2022-12-03 08:01:54,929] run_release_test.py: 160  Release test pipeline for test long_running_many_drivers completed. Returning with exit code = 0
 (e) py38   cherry-pick-30841  ~  ray 
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  cluster_url:   https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_rhdfm4es3cy7cjqxrpf3rmxr

[INFO 2022-12-03 08:01:46,158] log.py: 36  Observed the following results:

  iteration = 2716
  iteration_time = 61.77799606323242
  absolute_time = 1670054447.3899562
  elapsed_time = 86365.42208147049
  last_update = 1670054447.389978
  last_update_diff = 57.182602643966675

[INFO 2022-12-03 08:01:46,158] run_release_test.py: 160  Release test pipeline for test long_running_many_drivers completed. Returning with exit code = 0
 (e) py38  ~ 
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  wheels_url:    https://ray-ci-artifact-pr-public.s3.amazonaws.com/893c05b91cfbd9872a41e62f606bbe1cc90396fa/tmp/artifacts/.whl/ray-3.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
  cluster_url:   https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_rmjr4ysld7u28v84ywz16g6m

[INFO 2022-12-03 08:01:56,740] log.py: 36  Observed the following results:

  iteration = 2683
  iteration_time = 40.79775094985962
  absolute_time = 1670054456.1923652
  elapsed_time = 86367.7089073658
  last_update = 1670054456.1923857
  last_update_diff = 55.198883295059204

[INFO 2022-12-03 08:01:56,740] run_release_test.py: 160  Release test pipeline for test long_running_many_drivers completed. Returning with exit code = 0
 (e) py38  ~ 
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  cluster_url:   https://console.anyscale.com/o/anyscale-internal/projects/prj_FKRmeV5pA6X72aVscFALNC32/clusters/ses_vny6pic518euzr9zse23fg8m

[INFO 2022-12-03 08:02:17,735] log.py: 36  Observed the following results:

  iteration = 2739
  iteration_time = 34.63359594345093
  absolute_time = 1670054481.7758496
  elapsed_time = 86370.56573414803
  last_update = 1670054481.7758725
  last_update_diff = 54.35917401313782

[INFO 2022-12-03 08:02:17,736] run_release_test.py: 160  Release test pipeline for test long_running_many_drivers completed. Returning with exit code = 0
```

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -608,6 +608,10 @@ void CoreWorker::Shutdown() {
   if (io_thread_.joinable()) {
     io_thread_.join();
   }
+
+  // Shutdown gRPC server
+  core_worker_server_->Shutdown();
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:613]: oh interesting... so we've never shutdown the server so far?

**Review by rkooo567 — APPROVED** [review]
so this was a bug from before? can you tell me a bit more about the fix?

**Comment by iycheng** [comment]
@rkooo567 I'm not sure why it's becoming more frequently with the optimization. 
But the root cause is that we don't close the core worker and fail to do this will fail to drain the gRPC request and stop accepting the new ones.

Here the core worker is handling the pubsub request and will call gcs client to send something. This is the reason for the seg.

The fix will stop the core worker server so that it won't process the pubsub requests and thus fixing it.

**Pull request merged**